diff --git a/bfd/config.bfd b/bfd/config.bfd
index 30087e3b..3b432b82 100644
--- a/bfd/config.bfd
+++ b/bfd/config.bfd
@@ -568,6 +568,18 @@ case "${targ}" in
     targ_selvecs="x86_64_elf64_sol2_vec l1om_elf64_vec k1om_elf64_vec iamcu_elf32_vec i386_coff_vec i386_pei_vec x86_64_pe_vec x86_64_pei_vec"
     want64=true
     ;;
+#endif
+  i[3-7]86-*-hephaistos*)
+    targ_defvec=i386_elf32_vec
+    targ_selvecs=
+    targ64_selvecs=x86_64_elf64_vec
+    ;;
+#ifdef BFD64
+  x86_64-*-hephaistos*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-nto*)
     targ_defvec=i386_elf32_vec
diff --git a/config.sub b/config.sub
index 7384e919..f4687e4a 100755
--- a/config.sub
+++ b/config.sub
@@ -1705,7 +1705,7 @@ case $os in
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
 	     | os9* | macos* | osx* | ios* \
-	     | mpw* | magic* | mmixware* | mon960* | lnews* \
+	     | mpw* | magic* | mmixware* | mon960* | lnews* | hephaistos* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
 	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
diff --git a/gas/configure.tgt b/gas/configure.tgt
index 338892ad..33ff26ca 100644
--- a/gas/configure.tgt
+++ b/gas/configure.tgt
@@ -222,6 +222,7 @@ case ${generic_target} in
   i386-*-elfiamcu)			fmt=elf arch=iamcu ;;
   i386-*-elf*)				fmt=elf ;;
   i386-*-fuchsia*)			fmt=elf ;;
+  i386-*-hephaistos*)       fmt=elf ;;
   i386-*-genode*)			fmt=elf ;;
   i386-*-bsd*)				fmt=aout em=386bsd ;;
   i386-*-netbsdpe*)			fmt=coff em=pe ;;
diff --git a/ld/Makefile.am b/ld/Makefile.am
index f8e99325..37c4fb2b 100644
--- a/ld/Makefile.am
+++ b/ld/Makefile.am
@@ -281,6 +281,7 @@ ALL_EMULATION_SOURCES = \
 	eelf_i386_fbsd.c \
 	eelf_i386_ldso.c \
 	eelf_i386_sol2.c \
+    eelf_i386_hephaistos.c \
 	eelf_i386_vxworks.c \
 	eelf_iamcu.c \
 	eelf_s390.c \
@@ -453,6 +454,7 @@ ALL_64_EMULATION_SOURCES = \
 	eelf_k1om.c \
 	eelf_k1om_fbsd.c \
 	eelf_x86_64.c \
+    eelf_x86_64_hephaistos.c \
 	eelf_x86_64_cloudabi.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_sol2.c \
diff --git a/ld/Makefile.in b/ld/Makefile.in
index ef2e99e0..e8098f6e 100644
--- a/ld/Makefile.in
+++ b/ld/Makefile.in
@@ -771,6 +771,7 @@ ALL_EMULATION_SOURCES = \
 	eelf_i386_fbsd.c \
 	eelf_i386_ldso.c \
 	eelf_i386_sol2.c \
+	eelf_i386_hephaistos.c \
 	eelf_i386_vxworks.c \
 	eelf_iamcu.c \
 	eelf_s390.c \
@@ -942,6 +943,7 @@ ALL_64_EMULATION_SOURCES = \
 	eelf_k1om.c \
 	eelf_k1om_fbsd.c \
 	eelf_x86_64.c \
+    eelf_x86_64_hephaistos.c \
 	eelf_x86_64_cloudabi.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_sol2.c \
diff --git a/ld/configure.tgt b/ld/configure.tgt
index 6205d7c9..ae4f067c 100644
--- a/ld/configure.tgt
+++ b/ld/configure.tgt
@@ -293,6 +293,15 @@ hppa*-*-netbsd*)	targ_emul=hppanbsd
 			;;
 hppa*-*-openbsd*)	targ_emul=hppaobsd
 			;;
+i[3-7]86-*-hephaistos*)
+			targ_emul=elf_i386_hephaistos
+			targ_extra_emuls=elf_i386
+			targ64_extra_emuls="elf_x86_64_hephaistos elf_x86_64"
+			;;
+x86_64-*-hephaistos*)
+			targ_emul=elf_x86_64_hephaistos
+			targ_extra_emuls="elf_i386_hephaistos elf_x86_64 elf_i386"
+			;;
 i[3-7]86-*-nto-qnx*)	targ_emul=i386nto
 			;;
 i[3-7]86-*-go32)	targ_emul=i386go32
diff --git a/ld/emulparams/elf_i386_hephaistos.sh b/ld/emulparams/elf_i386_hephaistos.sh
new file mode 100644
index 00000000..b2f038cd
--- /dev/null
+++ b/ld/emulparams/elf_i386_hephaistos.sh
@@ -0,0 +1,4 @@
+. ${srcdir}/emulparams/elf_i386.sh
+GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
+TEXT_START_ADDR=0x40000000
\ No newline at end of file
diff --git a/ld/emulparams/elf_x86_64_hephaistos.sh b/ld/emulparams/elf_x86_64_hephaistos.sh
new file mode 100644
index 00000000..ba338a7c
--- /dev/null
+++ b/ld/emulparams/elf_x86_64_hephaistos.sh
@@ -0,0 +1,2 @@
+. ${srcdir}/emulparams/elf_x86_64.sh
+TEXT_START_ADDR=0x40000000
\ No newline at end of file
-- 
2.34.1