diff --git a/config.sub b/config.sub
index 38f3d037a..d2dac1f40 100644
--- a/config.sub
+++ b/config.sub
@@ -1724,7 +1724,7 @@ case $os in
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
 	     | os9* | macos* | osx* | ios* \
-	     | mpw* | magic* | mmixware* | mon960* | lnews* \
+	     | mpw* | magic* | mmixware* | mon960* | lnews* | hephaestos* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
 	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
diff --git a/fixincludes/mkfixinc.sh b/fixincludes/mkfixinc.sh
index df90720b7..e94e3e483 100644
--- a/fixincludes/mkfixinc.sh
+++ b/fixincludes/mkfixinc.sh
@@ -11,6 +11,7 @@ target=fixinc.sh

 # Check for special fix rules for particular targets
 case $machine in
+    *-hephaestos* | \
     i?86-*-cygwin* | \
     i?86-*-mingw32* | \
     x86_64-*-mingw32* | \
diff --git a/gcc/config.gcc b/gcc/config.gcc
index 95c91ee02..4485838b3 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -722,6 +722,12 @@ x86_cpus="generic intel"

 # Common parts for widely ported systems.
 case ${target} in
+*-*-hephaestos*)
+  gas=yes
+  gnu_ld=yes
+  default_use_cxa_atexit=yes
+  use_gcc_stdint=provide
+  ;;
 *-*-darwin*)
   tmake_file="t-darwin "
   tm_file="${tm_file} darwin.h"
@@ -1189,6 +1195,12 @@ case ${target} in
 esac

 case ${target} in
+i[34567]86-*-hephaestos*)
+    tm_file="${tm_file} i386/unix.h i386/att.h elfos.h glibc-stdint.h i386/i386elf.h hephaestos.h"
+    ;;
+x86_64-*-hephaestos*)
+    tm_file="${tm_file} i386/unix.h i386/att.h elfos.h glibc-stdint.h i386/i386elf.h i386/x86-64.h hephaestos.h"
+    ;;
 aarch64*-*-elf | aarch64*-*-fuchsia* | aarch64*-*-rtems*)
 	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h"
 	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-errata.h aarch64/aarch64-elf-raw.h"
diff --git a/gcc/config/hephaestos.h b/gcc/config/hephaestos.h
new file mode 100644
index 000000000..040834c99
--- /dev/null
+++ b/gcc/config/hephaestos.h
@@ -0,0 +1,20 @@
+/* Useful if you wish to make target-specific GCC changes. */
+#undef TARGET_hephaestos
+#define TARGET_hephaestos 1
+
+/* Default arguments you want when running your
+   i386-hephaestos-gcc/x86_64-hephaestos-gcc toolchain */
+#undef LIB_SPEC
+#define LIB_SPEC "-lc" /* link against c library */
+
+/* Additional predefined macros. */
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()        \
+    do                                  \
+    {                                   \
+        builtin_define("__hephaestos__");    \
+        builtin_define("__unix__");     \
+        builtin_assert("system=hephaestos"); \
+        builtin_assert("system=unix");  \
+        builtin_assert("system=posix"); \
+    } while (0);
diff --git a/libgcc/config.host b/libgcc/config.host
index e75a7af64..8a4f741cc 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -1569,6 +1569,14 @@ nvptx-*)
 	tmake_file="$tmake_file nvptx/t-nvptx"
 	extra_parts="crt0.o"
 	;;
+i[34567]86-*-hephaestos*)
+	extra_parts="$extra_parts crti.o crtbegin.o crtend.o crtn.o"
+	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
+	;;
+x86_64-*-hephaestos*)
+	extra_parts="$extra_parts crti.o crtbegin.o crtend.o crtn.o"
+	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
+	;;
 *)
 	echo "*** Configuration ${host} not supported" 1>&2
 	exit 1
diff --git a/libgcc/libgcov.h b/libgcc/libgcov.h
index 9d83a27ed..640c72425 100644
--- a/libgcc/libgcov.h
+++ b/libgcc/libgcov.h
@@ -189,6 +189,8 @@ extern struct gcov_info *gcov_list;
 #endif

 #include "gcov-io.h"
+//#include <stdlib.h>
+//#include "stdint.h"

 /* Structures embedded in coveraged program.  The structures generated
    by write_profile must match these.  */
diff --git a/libstdc++-v3/crossconfig.m4 b/libstdc++-v3/crossconfig.m4
index b3269cb88..2288cc941 100644
--- a/libstdc++-v3/crossconfig.m4
+++ b/libstdc++-v3/crossconfig.m4
@@ -5,6 +5,13 @@ dnl
 AC_DEFUN([GLIBCXX_CROSSCONFIG],[
 # Base decisions on target environment.
 case "${host}" in
+  *-hephaestos*)
+    GLIBCXX_CHECK_COMPILER_FEATURES
+    GLIBCXX_CHECK_LINKER_FEATURES
+    GLIBCXX_CHECK_MATH_SUPPORT
+    GLIBCXX_CHECK_STDLIB_SUPPORT
+    ;;
+
   arm*-*-symbianelf*)
     # This is a freestanding configuration; there is nothing to do here.
     ;;
